src/erlcloud_mms.erl

```diff
@@ -0,0 +1,257 @@
+%%% Request
+%%%------------------------------------------------------------------------------
+
+-type operation() :: string().
+-type json_return() :: {ok, json_term()} | {error, term()}.
+
+-spec request(aws_config(), operation(), json_term()) -> json_return().
+request(Config, Operation, Json) ->
+    case erlcloud_aws:update_config(Config) of
+        {ok, Config2} ->
+            request2(Config2, Operation, Json);
+        {error, _} = Error ->
+            Error
+    end.
+
+-spec request2(aws_config(), operation(), json_term()) -> json_return().
+request2(Config, Operation, Json) ->
+    Body = case Json of
+               [] -> <<"{}">>;
+               _ -> jsx:encode(Json)
+           end,
+    Headers = headers(Config, Operation, Body),
+    Request = #aws_request{service = mms,
+                           uri = uri(Config),
+                           method = post,
+                           request_headers = Headers,
+                           request_body = Body},
+    request_to_return(erlcloud_retry:request(Config, Request, fun result_fun/1)).
+
+-spec request_to_return(#aws_request{}) -> json_return().
+request_to_return(#aws_request{response_type = ok,
+                               response_body = Body}) ->
+    %% TODO check crc
+    {ok, jsx:decode(Body)};
+request_to_return(#aws_request{response_type = error,
+                               error_type = aws,
+                               httpc_error_reason = undefined,
+                               response_status = Status,
+                               response_status_line = StatusLine,
+                               response_body = Body}) ->
+    {error, {http_error, Status, StatusLine, Body}};
+request_to_return(#aws_request{response_type = error,
+                               error_type = aws,
+                               httpc_error_reason = Reason}) ->
+    {error, Reason};
+request_to_return(#aws_request{response_type = error,
+                               error_type = httpc,
+                               httpc_error_reason = Reason}) ->
+    {error, Reason}.
+
+-spec result_fun(#aws_request{}) -> #aws_request{}.
+result_fun(#aws_request{response_type = ok} = Request) ->
+    Request;
+result_fun(#aws_request{response_type = error,
+                        error_type = aws,
+                        response_status = Status} = Request) ->
+    %% == IMPLEMENTATION NOTES ==
+    %% Set `httpc_error_reason` to `undefined` in case of retry
+    Request2 = Request#aws_request{httpc_error_reason = undefined},
+    if
+        Status >= 400 andalso Status < 500 ->
+            client_error(Request2);
+        Status >= 500 ->
+            Request2#aws_request{should_retry = true};
+        Status < 400 ->
+            Request2#aws_request{should_retry = false}
+    end.
+
+-spec client_error(#aws_request{}) -> #aws_request{}.
+client_error(#aws_request{response_body = Body} = Request) ->
```

Comment:
i really do not like the idea of copy pasting ddb& ddb_streams code here. 
let's generalise it and use one copy